src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -12,34 +12,73 @@ namespace Microsoft.PowerShell.Commands
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private string _extension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension
+        {
+            get { return _extension; }
+            set { _extension = value; }
+        }
+
         /// <summary>
         /// returns a TemporaryFile
         /// </summary>
         protected override void EndProcessing()
         {
             string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    filePath = Path.GetTempFileName();
+                    try
+                    {
+                        string fileName = Path.GetRandomFileName();
+                        fileName = Path.ChangeExtension(fileName, _extension);
+                        filePath = Path.Combine(tempPath, fileName);
+                        using (new FileStream(filePath, FileMode.CreateNew)) { }
+                        creationOfFileSuccessful = true;
+                        WriteVerbose("Created temporary file {filePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                    }
+                    catch (Exception exception) // fatal error, which could be e.g. insufficient permissions, etc.
+                    {
+                        WriteError(
+                        new ErrorRecord(
+                                exception,
+                                "NewTemporaryFileWriteError",
+                                ErrorCategory.WriteError,
+                                tempPath));
+                        return;
+                    }
                 }
-                catch (Exception e)
+
+                if (!creationOfFileSuccessful)
                 {
                     WriteError(
                         new ErrorRecord(
-                            e,
-                            "NewTemporaryFileWriteError",
-                            ErrorCategory.WriteError,
-                            tempPath));
+                            new IOException("Could not find an available temporary file name in {tempPath}."),
```

Comment:
This exception message string should go into a resx file so it can be localized.  I think this would be the right file: Microsoft.PowerShell.Commands.Utility\resources\UtilityCommonStrings.resx